File: 01_Python/02_LeetCode/0001_TwoSum.py
"""
Problem description
 - https://leetcode.com/problems/two-sum/

Result
 - Runtime: 36 ms, faster than 99.75% of Python3 online submissions for Two Sum.
 - Memory Usage: 7.9 MB, less than 31.08% of Python3 online submissions for Two Sum.
"""


class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        index_dict = {}

        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in index_dict:
                return [i, index_dict[complement]]

            if nums[i] not in index_dict:
                index_dict[nums[i]] = i

File: 02_JavaScript/02_LeetCode/0001_TwoSum.js
/*
Problem description
  - https://leetcode.com/problems/two-sum/

Result
  - Runtime: 60 ms, faster than 94.01 % of JavaScript online submissions for Two Sum.
  - Memory Usage: 35.4 MB, less than 26.07 % of JavaScript online submissions for Two Sum.
*/


/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSum = function (nums, target) {
  let i = 0;
  const numsDict = {};
  while (i < nums.length) {
    const complement = target - nums[i];
    if (complement in numsDict) {
      return [i, numsDict[complement]];
    }

    if (!(nums[i] in numsDict)) {
      numsDict[nums[i]] = i;
    }

    i++;
  }
};